extensions/s3/src/main/java/io/deephaven/extensions/s3/S3OutputStream.java

```diff
@@ -0,0 +1,268 @@
+                .key(uri.key().orElseThrow())
+                .build();
+        // Note: We can add support for other parameters like tagging, storage class, encryption, permissions, etc. in
+        // future
+        final CompletableFuture<CreateMultipartUploadResponse> future =
+                s3AsyncClient.createMultipartUpload(createMultipartUploadRequest);
+        final CreateMultipartUploadResponse response;
+        try {
+            response = future.get();
+        } catch (final InterruptedException | ExecutionException e) {
+            throw handleS3Exception(e, String.format("initiating multipart upload for uri %s", uri), s3Instructions);
+        }
+        return response.uploadId();
+    }
+
+    /**
+     * Send a part request for the given buffer. This method assumes that the buffer is non-empty.
+     */
+    private void sendPartRequest(final OutgoingRequest request) throws IOException {
+        if (nextPartNumber > MAX_PART_NUMBER) {
+            throw new IOException("Cannot upload more than " + MAX_PART_NUMBER + " parts for uri " + uri + ", please" +
+                    " try again with a larger part size");
+        }
+        if (request.future != null) {
+            throw new IllegalStateException("Request already in progress for uri " + uri + " with part number " +
+                    nextPartNumber);
+        }
+        final UploadPartRequest uploadPartRequest = UploadPartRequest.builder()
+                .bucket(uri.bucket().orElseThrow())
+                .key(uri.key().orElseThrow())
+                .uploadId(uploadId)
+                .partNumber(nextPartNumber)
+                .build();
+        request.buffer.flip();
+        request.future = s3AsyncClient.uploadPart(uploadPartRequest,
+                AsyncRequestBody.fromByteBufferUnsafe(request.buffer));
+        request.partNumber = nextPartNumber;
+        nextPartNumber++;
+    }
+
+    private void waitForCompletion(final OutgoingRequest request) throws IOException {
+        final UploadPartResponse uploadPartResponse;
+        try {
+            uploadPartResponse = request.future.get();
+        } catch (final InterruptedException | ExecutionException e) {
+            throw handleS3Exception(e, String.format("waiting for part %d for uri %s to complete uploading",
+                    request.partNumber, uri), s3Instructions);
+        }
+        completedParts.add(SdkPojoConversionUtils.toCompletedPart(uploadPartResponse, request.partNumber));
+        request.buffer.clear();
+        request.future = null;
+        request.partNumber = INVALID_PART_NUMBER;
+    }
+
+    private void completeMultipartUpload() throws IOException {
+        // Complete all pending requests in the exact order they were sent
+        for (int partNumber = completedParts.size() + 1; partNumber < nextPartNumber; partNumber++) {
+            final OutgoingRequest request = pendingRequests.get((partNumber - 1) % numConcurrentParts);
```

Comment:
Please only change this if it makes more sense to you this way.
```suggestion
        final int partCount = nextPartNumber - 1;
        for (int partNumber = completedParts.size() + 1; partNumber <= partCount; partNumber++) {
            final OutgoingRequest request = pendingRequests.get((partNumber - 1) % numConcurrentParts);
```